routing/endpointregistry.go

```diff
@@ -82,58 +88,49 @@ func NewEndpointRegistry(o RegistryOptions) *EndpointRegistry {
 		o.LastSeenTimeout = defaultLastSeenTimeout
 	}
 
-	return &EndpointRegistry{
-		data:            map[string]*entry{},
-		lastSeen:        map[string]time.Time{},
+	result := &EndpointRegistry{
+		data:            sync.Map{},
 		lastSeenTimeout: o.LastSeenTimeout,
 		now:             time.Now,
 	}
+
+	for host, detected := range o.Detected {
+		e := &entry{}
+		e.detected.Store(detected)
+		e.lastSeen.Store(result.now())
+		result.data.Store(host, e)
+	}
+
+	return result
 }
 
 func (r *EndpointRegistry) GetMetrics(key string) Metrics {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+	e, _ := r.data.LoadOrStore(key, newEntry())
+	old := e.(*entry)
 
-	e := r.getOrInitEntryLocked(key)
-	copy := &entry{}
-	*copy = *e
+	copy := newEntry()
+	copy.detected.Store(old.detected.Load())
+	atomic.StoreInt64(&copy.inflightRequests, atomic.LoadInt64(&old.inflightRequests))
+	// no need to copy lastSeen, because it is needed only for endpointregistry itself
 	return copy
 }
 
-func (r *EndpointRegistry) SetDetectedTime(key string, detected time.Time) {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+func (r *EndpointRegistry) setDetectedTime(key string, detected time.Time) {
+	e, _ := r.data.LoadOrStore(key, newEntry())
+	e.(*entry).detected.Store(detected)
```

Comment:
I think I will change this command to compare-and-swap, it seems to fit better here.